### PR TITLE
Documentation for sea-orm

### PR DESCRIPTION
**Pull request by charleschege**
Adds documentation for all public types for this crate.

### CODE REVIEW

**Comment by tyt2y3** [comment]
@billy1624 know why the build failed?

**Comment by charleschege** [comment]
@tyt2y3 I am investigating a way to solve the issue. What I have discovered on my end is that the Github workflow runs `cargo test --workspace` which tests even the documentation.  

Somehow the test is not detecting crate imports in the documentation. There are errors like
```
  |
3 | use sea_orm::entity::prelude::*;
  |     ^^^^^^^ use of undeclared crate or module `sea_orm`

```
even though `use sea_orm::entity::prelude::*; ` is imported.

**Comment by billy1624** [comment]
> @billy1624 know why the build failed?

Which error? There are a lot of errors loll

**Comment by charleschege** [comment]
@billy1624 @tyt2y3 Can someone help me with this issue?

Running `cargo test --workspace` seems to be the issue. It is not detecting `sea_orm_macros` in documentation examples.

But when i run `cargo test --doc` or `cargo test --all-targets` everything works fine, the macros and modules are detected. Does anyone have suggestions on how to import macros on documentation tests ?

**Comment by billy1624** [comment]
Click on the line number on the left and copy & paste the URL

image

**Comment by billy1624** [comment]
> @billy1624 @tyt2y3 Can someone help me with this issue?
> 
> Running `cargo test --workspace` seems to be the issue. It is not detecting `sea_orm_macros` in documentation examples.
> 
> But when i run `cargo test --doc` or `cargo test --all-targets` everything works fine, the macros and modules are detected. Does anyone have suggestions on how to import macros on documentation tests ?

Is this related? #249

**Comment by billy1624** [comment]
Hey @charleschege, is it possible for you to grant write access to us? I want to add the follow code snippet to `.github/workflows/rust.yml`.

```yml
  test-temp:
    name: Unit Test (Temp)
    runs-on: ubuntu-20.04
    strategy:
      matrix:
        arg: ["--workspace", "--doc", "--all-targets"]
    steps:
      - uses: actions/checkout@v2

      - uses: actions-rs/toolchain@v1
        with:
          profile: minimal
          toolchain: stable
          override: true

      - uses: actions-rs/cargo@v1
        with:
          command: test
          args: ${{ matrix.args }}
```

https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by charleschege** [comment]
I have added access to the repo

**Comment by charleschege** [comment]
> Hey @charleschege, can you point out which error should I look into? E.g. https://github.com/SeaQL/sea-orm/runs/4054619075?check_suite_focus=true#step:4:589 this error. Thanks!

My issues stem from this test run:
https://github.com/SeaQL/sea-orm/runs/4054619075?check_suite_focus=true

**Comment by billy1624** [comment]
Hey @charleschege, it’s already 11pm (HKT) for me. I’m about to sleep. Will talk to you tmr.

File: src/entity/active_model.rs
Comment (jam1garner) [inline src/entity/active_model.rs:27]: Could some documentation be added covering why this takes an unused `Option<bool>`? I don't feel the type's purpose is very clear, even having dug through the source (off-topic for this PR, but would this perhaps be better served by either a 3-value enum or an `Option` of a 2-value enum? that would allow this to be more self-documenting)

Comment (charleschege) [reply src/entity/active_model.rs:48]: Documenting that is a little bit tricky since the field of `ActiveValue` are private. I think a user would have to build the documentation with `--document-private-items` set to see any inner struct documentation.

I have tried to provide an overview of the internals of `ActiveValue`and a code snippet in the documentation. Take a look at commit 30814f0053bf0cfe6a8a0001cf63d31600926802

Comment (tyt2y3) [reply src/entity/active_model.rs:48]: Hi there, welcome! You have made a good point.

It certainly is trying to mock a enum, where `Set` & `Unset` etc is like `Ok` and `Err`. It could as well be `Unset(())`.

I was waiting for private enum variant https://github.com/rust-lang/rfcs/pull/2028#issuecomment-308241043 to hide the `Unchanged` variant from public use.

And that's why.

Comment (tyt2y3) [reply src/entity/active_model.rs:48]: > I have tried to provide an overview of the internals of `ActiveValue`and a code snippet in the documentation. Take a look at commit [30814f0](https://github.com/SeaQL/sea-orm/commit/30814f0053bf0cfe6a8a0001cf63d31600926802)

Thank you

File: sea-orm-macros/Cargo.toml
Comment (billy1624) [inline sea-orm-macros/Cargo.toml:27]: Hey @charleschege, I have fixed all the errors inside `sea-orm` doctest on fcf3ea9. And I know why `cargo test --doc` passes while `cargo test --workspace` fails.

In `sea-orm-macros/src/lib.rs`, you have include some code snippets that requires `sea-orm`. However, `sea-orm-macros` does not depends on `sea-orm`. So, all of the doctests failed. By adding `dev-dependencies` will solve the problem. 6904b9f

File: sea-orm-macros/src/lib.rs
Comment (billy1624) [inline sea-orm-macros/src/lib.rs:24]: Btw... for some derive macros it's quite complicated to show its usage, as it requires implementing other traits manually or using other derive macros.

I was trying to make Ferris compile for us by adding some placeholder structs, enums and impls. So you can feel free to revert any changes that I made :)

**Review by billy1624 — CHANGES_REQUESTED** [review]

**Comment by tyt2y3** [comment]
wow now it builds! a big thanks to @billy1624

File: .github/workflows/rust.yml
Comment (billy1624) [inline .github/workflows/rust.yml:282]: Just a kind reminder. Removing it before merge :P

**Review by billy1624 — CHANGES_REQUESTED** [review]

File: src/entity/active_model.rs
Comment (jam1garner) [reply src/entity/active_model.rs:48]: in the case of

```rust
fn Unset<V>(_: Option<bool>)
//         ^^^^^^^^^^^^^^^
```

Why the unused parameter? I understand that Unset is supposed to emulate a variant of an enum but that still doesn't explain to me why the parameter.

File: sea-orm-macros/Cargo.toml
Comment (charleschege) [reply sea-orm-macros/Cargo.toml:27]: Thanks! This has resolved all the issues I had when testing documentation examples.

File: sea-orm-macros/src/lib.rs
Comment (charleschege) [reply sea-orm-macros/src/lib.rs:24]: Thanks! That improves the documentation further.

**Review by charleschege — COMMENTED** [review]
LGTM

File: src/entity/active_model.rs
Comment (tyt2y3) [reply src/entity/active_model.rs:48]: Because `Unset()` is not a valid enum. I agree this is a little confusing. May be I should change it in `0.5` to something like

```rust
enum ActiveValue {
    Set(Value),
    UnchangedDoNotUse,
    NotSet,    
}
```

**Review by tyt2y3 — APPROVED** [review]
Thank you for the meticulous work

File: src/entity/active_model.rs
Comment (jam1garner) [reply src/entity/active_model.rs:48]: @tyt2y3 something like that would be a positive change, yes. Tbh if the argument has no meaning, the name `Unset` seems fine to me (albeit slightly ambiguous if you mean "unset" (verb, action, to switch to a no longer set state)  or "unset" (adjective, state, not currently set)), I mostly think just getting rid of the argument is enough for it to make sense to me. Thanks!

**Pull request merged**
